File: gen_node_docs.py
import re
import yaml
from pathlib import Path

# Paths to your directories
nodes_dir = Path("../PixiEditor/src/PixiEditor.ChangeableDocument/Changeables/Graph/Nodes")
viewmodels_dir = Path("../PixiEditor/src/PixiEditor/ViewModels/Document/Nodes")
output_dir = Path("src/content/docs/usage/Node Graph/Nodes/")

# Match ViewModel metadata
viewmodel_pattern = re.compile(r'\[NodeViewModel\("([^"]+)",\s*"([^"]+)",\s*([^\)]+)\)\]')
pair_pattern = re.compile(r'\[PairNode\(typeof\(([^)]+)\)')
input_pattern_generic = re.compile(
    r'Create(?:(?P<kind>Func))?Input<([^>]+)>\(([^,]+),\s*"([^"]+)"(?:,\s*(.+?))?\)'
)

output_pattern_generic = re.compile(
    r'Create(?:(?P<kind>Func))?Output<([^>]+)>\(([^,]+),\s*"([^"]+)"(?:,\s*(.+?))?\)'
)

# Non-generic inputs/outputs
input_pattern_nongeneric = re.compile(
    r'Create(?:(?P<kind>Func|Render|SyncedType))?Input\(([^,]+),\s*"([^"]+)"(?:,\s*(.+?))?\)'
)

output_pattern_nongeneric = re.compile(
    r'Create(?:(?P<kind>Func|Render|SyncedType))?Output\(([^,]+),\s*"([^"]+)"(?:,\s*(.+?))?\)'
)

property_pattern = re.compile(
    r'public\s+(InputProperty|FuncOutputProperty|OutputProperty|FuncInputProperty)<([^>]+)>\s+(\w+)\s*\{'
)

def screaming_to_words(text):
    return " ".join(word.capitalize() for word in text.split("_"))

def extract_viewmodel_metadata(file_content):
    match = viewmodel_pattern.search(file_content)
    if not match:
        return None
    name, category, icon = match.groups()
    return {
        "name": screaming_to_words(name),
        "category": screaming_to_words(category),
        "icon": f"icon-{icon.split('.')[-1].lower()}"
    }

def map_type(type_name):
    type_mapping = {
        "Float1": "Double",
        "Float2": "VecD",
        "Int1": "Integer",
        "Half3": "Vec3D",
        "Half4": "Color",
        "Int2": "VecI",
    }
    return type_mapping.get(type_name, type_name)

def extract_property_types(file_content):
    prop_types = {}
    for match in property_pattern.finditer(file_content):
        _, generic_type, prop_name = match.groups()
        prop_types[prop_name] = generic_type.strip()
    return prop_types

def extract_inputs_outputs_from_content(file_content, prop_types):
    inputs = []
    outputs = []

    # generic inputs
    for match in input_pattern_generic.finditer(file_content):
        kind, type_, name, label, default = match.groups()
        inputs.append({
            "name": screaming_to_words(label),
            "type": map_type(type_.strip()),
            "description": "TODO: Add a description.",
            "isContextful": kind == "Func",
            "default": default.strip() if default else None
        })

    # non-generic inputs
    for match in input_pattern_nongeneric.finditer(file_content):
        kind, name, label, default = match.groups()
        type_ = prop_types.get(name, "unknown")
        inputs.append({
            "name": screaming_to_words(label),
            "type": map_type(type_),
            "description": "TODO: Add a description.",
            "isContextful": kind == "Func",
            "default": default.strip() if default else None
        })

    # generic outputs
    for match in output_pattern_generic.finditer(file_content):
        kind, type_, name, label, default = match.groups()
        outputs.append({
            "name": screaming_to_words(label),
            "type": map_type(type_.strip()),
            "description": "TODO: Add a description.",
            "isContextful": kind == "Func",
            "default": default.strip() if default else None
        })

    # non-generic outputs
    for match in output_pattern_nongeneric.finditer(file_content):
        kind, name, label, default = match.groups()
        type_ = prop_types.get(name, "unknown")
        outputs.append({
            "name": screaming_to_words(label),
            "type": map_type(type_),
            "description": "TODO: Add a description.",
            "isContextful": kind == "Func",
            "default": default.strip() if default else None
        })

    return inputs, outputs

def find_base_class_name(file_content):
    match = re.search(r'class\s+\w+\s*:\s*([\w\d_]+)', file_content)
    if match:
        return match.group(1)
    return None

def find_node_file_by_classname(classname):
    for file in nodes_dir.rglob(f"{classname}Node.cs"):
        return file
    for file in nodes_dir.rglob(f"{classname}.cs"):
        return file
    return None

def extract_node_metadata_recursive(file_path, collected_inputs=None, collected_outputs=None, collected_isPair=False, collected_hasPreview=False):
    if collected_inputs is None:
        collected_inputs = []
    if collected_outputs is None:
        collected_outputs = []

    content = file_path.read_text(encoding="utf-8")
    prop_types = extract_property_types(content)

    inputs, outputs = extract_inputs_outputs_from_content(content, prop_types)
    collected_inputs.extend(inputs)
    collected_outputs.extend(outputs)

    if not collected_isPair:
        collected_isPair = bool(pair_pattern.search(content))
    if not collected_hasPreview:
        collected_hasPreview = "RenderNode" in content or "IPreviewRenderable" in content

    base_class = find_base_class_name(content)
    if not base_class or base_class == "Node":
        return {
            "inputs": collected_inputs,
            "outputs": collected_outputs,
            "isPair": collected_isPair,
            "hasPreview": collected_hasPreview
        }

    base_file = find_node_file_by_classname(base_class)
    if base_file is None:
        return {
            "inputs": collected_inputs,
            "outputs": collected_outputs,
            "isPair": collected_isPair,
            "hasPreview": collected_hasPreview
        }

    return extract_node_metadata_recursive(base_file, collected_inputs, collected_outputs, collected_isPair, collected_hasPreview)

def to_kebab_case(name: str) -> str:
    s1 = re.sub('(.)([A-Z][a-z]+)', r'\1-\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1-\2', s1).lower()

def merge_metadata(viewmodel_meta, node_meta):
    return {
        "name": viewmodel_meta["name"].replace("Node", ""),
        "category": viewmodel_meta["category"],
        "icon": viewmodel_meta["icon"],
        "isPair": node_meta.get("isPair", False),
        "hasPreview": node_meta.get("hasPreview", False),
        "inputs": node_meta.get("inputs") or None,
        "outputs": node_meta.get("outputs") or None,
        "description": "TODO: Add a description."
    }

def write_mdx_file(metadata, name, existing_files):
    if f"{name}.mdx" in existing_files:
        print(f"Skipping {name}: File already exists somewhere in output_dir.")
        return

    title = metadata['name']
    frontmatter = {
        "title": title,
        "node": metadata
    }

    yaml_frontmatter = yaml.dump(frontmatter, sort_keys=False, allow_unicode=True)
    content = f"---\n{yaml_frontmatter}---\n"

    output_dir.mkdir(parents=True, exist_ok=True)
    mdx_path = output_dir / f"{name}.mdx"

    with mdx_path.open("w", encoding="utf-8") as f:
        f.write(content)

def main():
    nodes = {f.name.replace("Node.cs", ""): f for f in nodes_dir.rglob("*Node.cs")}
    viewmodels = {f.name.replace("NodeViewModel.cs", ""): f for f in viewmodels_dir.rglob("*NodeViewModel.cs")}

    # 🔥 Precompute existing mdx files (recursive)
    existing_mdx_files = {f.name for f in output_dir.rglob("*.mdx")}

    for key in sorted(set(nodes.keys()) & set(viewmodels.keys())):
        viewmodel_path = viewmodels[key]
        node_path = nodes[key]

        vm_content = viewmodel_path.read_text(encoding="utf-8")

        viewmodel_meta = extract_viewmodel_metadata(vm_content)
        if not viewmodel_meta:
            print(f"Skipping {key}: No ViewModel metadata found.")
            continue

        node_meta = extract_node_metadata_recursive(node_path)
        merged = merge_metadata(viewmodel_meta, node_meta)

        write_mdx_file(merged, to_kebab_case(key), existing_mdx_files)

if __name__ == "__main__":
    main()

File: gen_node_docs.cs
#!/usr/bin/dotnet run
#:package YamlDotNet@18.1.0

using System.Reflection;
using System.Text;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

const string changeableDll = "../PixiEditor/src/PixiEditor.Desktop/bin/Debug/net8.0/PixiEditor.ChangeableDocument.dll";
const string editorDll = "../PixiEditor/src/PixiEditor.Desktop/bin/Debug/net8.0/PixiEditor.dll";
const string uiCommonDll = "../PixiEditor/src/PixiEditor.Desktop/bin/Debug/net8.0/PixiEditor.UI.Common.dll";

const string outputDir = "src/content/docs/usage/Node Graph/Nodes/";

var changeableAssembly = Assembly.LoadFrom(changeableDll);
var editorAssembly = Assembly.LoadFrom(editorDll);
var uiCommonAssembly = Assembly.LoadFrom(uiCommonDll);

var nodeType = changeableAssembly.GetType(
    "PixiEditor.ChangeableDocument.Changeables.Graph.Nodes.Node");

if (nodeType == null)
    throw new Exception("Node type not found");


var existingDocs = new HashSet<string>(
    StringComparer.OrdinalIgnoreCase);

if (Directory.Exists(outputDir))
{
    foreach (var file in Directory.GetFiles(
        outputDir,
        "*.mdx",
        SearchOption.AllDirectories))
    {
        // Check filename
        existingDocs.Add(
            Path.GetFileName(file));


        // Check PIXIEDITOR marker
        var lines = File.ReadLines(file)
            .Take(3)
            .ToArray();

        foreach (var line in lines)
        {
            if (line.StartsWith("#PIXIEDITOR:",
                    StringComparison.OrdinalIgnoreCase))
            {
                var nodeName = line["#PIXIEDITOR:".Length..]
                    .Trim();

                if (!string.IsNullOrWhiteSpace(nodeName))
                {
                    existingDocs.Add(
                        "#PIXIEDITOR:" + nodeName);
                }
            }
        }
    }
}


var nodeClasses = changeableAssembly
    .GetTypes()
    .Where(t =>
        !t.IsAbstract &&
        nodeType.IsAssignableFrom(t));


Directory.CreateDirectory(outputDir);


var serializer = new SerializerBuilder()
    .ConfigureDefaultValuesHandling(
        DefaultValuesHandling.OmitNull)
    .Build();



foreach (var node in nodeClasses.OrderBy(x => x.Name))
{
    var viewModel = FindViewModel(
        node,
        editorAssembly);

    if (viewModel == null)
        continue;


var viewModelMetadata =
    ExtractViewModelMetadata(
        viewModel,
        uiCommonAssembly);

    if (viewModelMetadata == null)
    {
        Console.WriteLine(
            $"Skipping {node.Name}: missing metadata");

        continue;
    }


    var nodeMetadata =
        ExtractNodeMetadata(node);


    var docName =
        ToKebabCase(
            node.Name.Replace("Node", ""));

if (existingDocs.Contains(docName + ".mdx"))
{
    Console.WriteLine(
        $"Skipping {node.Name}: {docName}.mdx already exists.");

    continue;
}

if (existingDocs.Contains("#PIXIEDITOR:" + node.Name.Replace("Node", "")))
{
    Console.WriteLine(
        $"Skipping {node.Name}: PIXIEDITOR marker exists.");

    continue;
}


    var metadata = new Dictionary<string, object>
    {
        ["title"] = viewModelMetadata["name"],

        ["node"] = new Dictionary<string, object>
        {
            ["name"] = viewModelMetadata["name"],
            ["category"] = viewModelMetadata["category"],
            ["icon"] = viewModelMetadata["icon"],

            ["isPair"] = nodeMetadata.IsPair,
            ["hasPreview"] = nodeMetadata.HasPreview,

            ["inputs"] =
                nodeMetadata.Inputs.Count > 0
                    ? nodeMetadata.Inputs
                    : null,

            ["outputs"] =
                nodeMetadata.Outputs.Count > 0
                    ? nodeMetadata.Outputs
                    : null,

            ["description"] =
                "TODO: Add a description."
        }
    };


    var yaml = serializer.Serialize(metadata);
    var sb = new StringBuilder();

sb.AppendLine("---");
sb.Append(yaml);
sb.AppendLine("---");

foreach (var enumType in nodeMetadata.Enums)
{
    sb.AppendLine();
    sb.AppendLine($"## {ScreamingToWords(enumType.Name)}");

    foreach (var value in Enum.GetNames(enumType))
    {
        sb.AppendLine();
        sb.AppendLine($"### {ScreamingToWords(value)}");
        sb.AppendLine();
        sb.AppendLine("TODO: Add description.");
    }
}

var category =
    viewModelMetadata["category"]
        .ToString()!;


var categoryFolder =
    Path.Combine(
        outputDir,
        category);


Directory.CreateDirectory(categoryFolder);


var path = Path.Combine(
    categoryFolder,
    docName + ".mdx");


    File.WriteAllText(
        path,
        sb.ToString(),
        Encoding.UTF8);


    Console.WriteLine(
        $"Generated {path}");
}



static Type? FindViewModel(
    Type node,
    Assembly assembly)
{
    return assembly
        .GetTypes()
        .FirstOrDefault(t =>
            t.Name == node.Name + "ViewModel");
}



static Dictionary<string, object>? ExtractViewModelMetadata(
    Type vm,
    Assembly editorAssembly)
{
    var attribute = vm
        .GetCustomAttributes()
        .FirstOrDefault(a =>
            a.GetType().Name ==
            "NodeViewModelAttribute");


    if (attribute == null)
        return null;


    var values = attribute
        .GetType()
        .GetProperties()
        .Select(p => p.GetValue(attribute))
        .ToArray();


    var name =
        values.ElementAtOrDefault(0)
            ?.ToString()
        ?? string.Empty;


    var category =
        values.ElementAtOrDefault(1)
            ?.ToString()
        ?? string.Empty;


    var iconValue =
        values.ElementAtOrDefault(2);


var icon =
    GetIconName(
        values.ElementAtOrDefault(2),
        editorAssembly);


    return new()
    {
        ["name"] =
            ScreamingToWords(
                name.Replace("_NODE", "")),

        ["category"] =
            ScreamingToWords(category),

        ["icon"] =
            icon
    };
}



static NodeMetadata ExtractNodeMetadata(Type node)
{
    var nodeMetadata = new NodeMetadata();

    foreach (var property in node.GetProperties(
        BindingFlags.Public |
        BindingFlags.Instance |
        BindingFlags.FlattenHierarchy))
    {
        var type = property.PropertyType;

        if (!type.IsGenericType)
            continue;

        var genericDefinition = type.GetGenericTypeDefinition();
        var genericName = genericDefinition.Name;

        var argumentType = type.GetGenericArguments()[0];
        var argument = argumentType.Name;

        var isContextful =
            genericName.StartsWith("FuncInputProperty") ||
            genericName.StartsWith("FuncOutputProperty");

        var port = new Dictionary<string, object>
        {
            ["name"] = ScreamingToWords(property.Name),
            ["type"] = MapType(argument),
            ["description"] = "TODO: Add a description."
        };

        if (isContextful)
            port["isContextful"] = true;

        if (argumentType.IsEnum)
        {
            port["typeLink"] = "#" + ToKebabCase(argumentType.Name);

            if (!nodeMetadata.Enums.Contains(argumentType))
                nodeMetadata.Enums.Add(argumentType);
        }

        if (genericName.EndsWith("InputProperty`1"))
        {
            nodeMetadata.Inputs.Add(port);
        }
        else if (genericName.EndsWith("OutputProperty`1"))
        {
            nodeMetadata.Outputs.Add(port);
        }
    }

    nodeMetadata.IsPair =
        node.GetCustomAttributes()
            .Any(a => a.GetType().Name == "PairNodeAttribute");

    nodeMetadata.HasPreview =
        node.GetInterfaces()
            .Any(i =>
                i.Name.Contains("Preview") ||
                i.Name.Contains("Renderable"));

    return nodeMetadata;
}

static string GetIconName(
    object? iconValue,
    Assembly assembly)
{
    if (iconValue == null)
        return string.Empty;


    var iconType = assembly
        .GetTypes()
        .FirstOrDefault(t =>
            t.Name == "PixiPerfectIcons");


    if (iconType == null)
        return string.Empty;


    var iconValueString = iconValue.ToString();


    var field = iconType
        .GetFields(
            BindingFlags.Public |
            BindingFlags.Static)
        .FirstOrDefault(f =>
            f.IsLiteral &&
            f.GetRawConstantValue()?.ToString()
                == iconValueString);


    if (field == null)
        return string.Empty;


    return "icon-" + ToKebabCase(field.Name);
}

static string MapType(
    string type)
{
    return type switch
    {
        "Float1" => "Double",
        "Float2" => "VecD",
        "Int1" => "Integer",
        "Half3" => "Vec3D",
        "Half4" => "Color",
        "Int2" => "VecI",
        "Int32" => "Integer",
        "Single" => "Float",
        _ => type
    };
}



static string ScreamingToWords(string? value)
{
    if (string.IsNullOrWhiteSpace(value))
        return string.Empty;

    value = System.Text.RegularExpressions.Regex.Replace(
        value,
        "([a-z0-9])([A-Z])",
        "$1 $2");

    value = System.Text.RegularExpressions.Regex.Replace(
        value,
        "([A-Z]+)([A-Z][a-z])",
        "$1 $2");

    value = value.Replace('_', ' ');

    return string.Join(
        " ",
        value
            .Split(
                ' ',
                StringSplitOptions.RemoveEmptyEntries)
            .Select(word =>
                char.ToUpperInvariant(word[0]) +
                word.Substring(1).ToLowerInvariant()));
}

static string ToKebabCase(string value)
{
    value = System.Text.RegularExpressions.Regex.Replace(
        value,
        "([a-z0-9])([A-Z])",
        "$1-$2");


    value = System.Text.RegularExpressions.Regex.Replace(
        value,
        "([A-Z]+)([A-Z][a-z])",
        "$1-$2");


    return value
        .Replace('_', '-')
        .ToLowerInvariant();
}

class NodeMetadata
{
    public List<object> Inputs { get; set; } = [];
    public List<object> Outputs { get; set; } = [];

    public List<Type> Enums { get; } = [];

    public bool IsPair { get; set; }
    public bool HasPreview { get; set; }
}